src/extensions/struct-conn/index.ts

```diff
@@ -0,0 +1,365 @@
+        { label: `${conn.id} (atoms)`, type: { name: 'expression', params: expressionByAtoms } },
+        { tags: [TAGS.ATOM_SEL] }
+    ).commit();
+
+    if (selectionByResidues.isOk) {
+        await plugin.build().to(selectionByResidues).apply(
+            StructureRepresentation3D,
+            RESIDUES_VISUAL_PARAMS,
+            { tags: [TAGS.RESIDUE_REPR] }
+        ).commit();
+    }
+    if (selectionByAtom.isOk) {
+        const atomBalls = await plugin.build().to(selectionByAtom).apply(
+            StructureRepresentation3D,
+            ATOMS_VISUAL_PARAMS,
+            { tags: [TAGS.ATOM_REPR] }
+        ).commit();
+        plugin.managers.camera.focusRenderObjects(atomBalls.data?.repr.renderObjects, { extraRadius: EXTRA_RADIUS });
+    }
+}
+
+/** Remove anything created by `addStructConnInspection` */
+async function removeAllStructConnInspections(plugin: PluginContext) {
+    const selNodes = [
+        ...plugin.state.data.selectQ(q => q.root.subtree().withTag(TAGS.RESIDUE_SEL)),
+        ...plugin.state.data.selectQ(q => q.root.subtree().withTag(TAGS.ATOM_SEL)),
+    ];
+    for (const node of selNodes) {
+        await plugin.build().delete(node).commit();
+    }
+}
+
+/** Hide all carbohydrate SNFG visuals */
+function hideSnfgNodes(plugin: PluginContext) {
+    const hiddenNodes = getExtensionState(plugin).hiddenNodes;
+    const snfgNodes = plugin.state.data.selectQ(q => q.root.subtree().withTag('branched-snfg-3d'));
+    for (const node of snfgNodes) {
+        if (!node.state.isHidden) {
+            setSubtreeVisibility(plugin.state.data, node.transform.ref, true); // true means hidden
+            hiddenNodes.set(node.transform.ref, true);
+        }
+    }
+}
+
+/** Make visible all carbohydrate SNFG visuals that have been hidden by `hideSnfgNodes` */
+function unhideSnfgNodes(plugin: PluginContext) {
+    const hiddenNodes = getExtensionState(plugin).hiddenNodes;
+    for (const nodeRef of Array.from(hiddenNodes.keys())) {
+        try {
+            setSubtreeVisibility(plugin.state.data, nodeRef, false); // false means visible
+        } catch {
+            // this is OK, the node has been removed
+        }
+    }
+    hiddenNodes.clear();
+}
+
+
+interface ExtensionState {
+    structConnCache: NodeMap<{ [id: string]: StructConnRecord }>,
+    hiddenNodes: NodeMap<boolean>,
+}
+function createExtensionState(plugin: PluginContext): ExtensionState {
+    return {
+        structConnCache: new NodeMap<{ [id: string]: StructConnRecord }>(plugin),
+        hiddenNodes: new NodeMap<boolean>(plugin),
+    };
+}
+function getExtensionState(plugin: PluginContext): ExtensionState {
+    return (plugin.customState as any).StructConn ??= createExtensionState(plugin);
+}
+
+
+/** Map in which keys are refs of plugin state tree nodes,
+ * and they get automatically deleted from the map
+ * when the node is deleted from the tree. */
+class NodeMap<T> extends PluginBehavior.WithSubscribers<undefined> {
```

Comment:
Why do you need this? It seems to me you can always query the current state tree to get all the info you need to change it. Extract struct conn can be cached directly at the structure level.

Reply:
This was to ensure that cached data that relate to a node are removed when that node is deleted.
I use in 2 places: for caching the parsed struct_conns, and for keeping track of the hidden SNFG nodes. The first could be probably moved to `model._staticPropertyData`. The latter I'm not sure about - it could cause weird behavior if removed (https://github.com/molstar/molstar/pull/802#discussion_r1179131514).

Reply:
I managed to get rid of the need for NodeMap and removed it. Thanks for help.